File: leetcode/DP/300.Longest-Increasing-Subsequence/300.Longest-Increasing-Subsequence.go
package leetcode

import "sort"

/* 300. 最长递增子序列
给你一个整数数组 nums ，找到其中最长严格递增子序列的长度。

子序列 是由数组派生而来的序列，删除（或不删除）数组中的元素而不改变
其余元素的顺序。例如，[3,6,2,7] 是数组 [0,3,1,6,2,2,7] 的子序列。

 
示例 1：
输入：nums = [10,9,2,5,3,7,101,18]
输出：4
解释：最长递增子序列是 [2,3,7,101]，因此长度为 4 。

示例 2：
输入：nums = [0,1,0,3,2,3]
输出：4

示例 3：
输入：nums = [7,7,7,7,7,7,7]
输出：1
*/

// 参考halfrost O(n^2) DP
func lengthOfLIS(nums []int) int {
	dp, res := make([]int, len(nums)+1), 0
	dp[0] = 0
	for i := 1; i <= len(nums); i++ {
		for j := 1; j < i; j++ {
			if nums[j-1] < nums[i-1] {
				dp[i] = max(dp[i], dp[j])
			}
		}
		dp[i] = dp[i] + 1
		res = max(res, dp[i])
	}
	return res
}

func max(a, b int) int {
	if a > b {
		return a
	}
	return b
}

// 解法二 O(n log n) DP
func lengthOfLIS_v1(nums []int) int {
	dp := []int{}
	for _, num := range nums {
		i := sort.SearchInts(dp, num)
		if i == len(dp) {
			dp = append(dp, num)
		} else {
			dp[i] = num
		}
	}
	return len(dp)
}

/*
解题思路

给定一个整数序列，求其中的最长上升子序列的长度。
这一题就是经典的最长上升子序列的问题。

dp[i] 代表为第 i 个数字为结尾的最长上升子序列的长度。

换种表述，dp[i] 代表 [0,i] 范围内，选择数字 nums[i] 可以获得的最长上升子序列的长度。
状态转移方程为 dp[i] = max( 1 + dp[j]) ，其中 j < i && nums[j] > nums[i]，
取所有满足条件的最大值。时间复杂度 O(n^2)

这道题还有一种更快的解法。考虑这样一个问题，我们是否能用一个数组，
记录上升子序列的最末尾的一个数字呢？如果这个数字越小，那么这个子序列往后面添加数字的几率就越大，
那么就越可能成为最长的上升子序列。

举个例子：nums = [4,5,6,3]，它的所有的上升子序列为
    len = 1   :      [4], [5], [6], [3]   => tails[0] = 3
    len = 2   :      [4, 5], [5, 6]       => tails[1] = 5
    len = 3   :      [4, 5, 6]            => tails[2] = 6
其中 tails[i] 中存储的是所有长度为 i + 1 的上升子序列中末尾最小的值。
也很容易证明 tails 数组里面的值一定是递增的(因为我们用末尾的数字描述最长递增子序列)。
既然 tails 是有序的，我们就可以用二分查找的方法去更新这个 tail 数组里面的值。
更新策略如下：(1). 如果 x 比所有的 tails 元素都要大，那么就直接放在末尾，并且 tails 数组长度加一；
(2). 如果 tails[i-1] < x <= tails[i]，则更新 tails[i]，因为 x 更小，更能获得最长上升子序列。
最终 tails 数组的长度即为最长的上升子序列。
这种做法的时间复杂度 O(n log n)。
*/

File: leetcode-cpp/300.Longest-Increasing-Subsequence/300.Longest-Increasing-Subsequence.cpp
/* 300. 最长递增子序列
给你一个整数数组 nums ，
找到其中最长严格递增子序列的长度。
子序列 是由数组派生而来的序列，删除（或不删除）
数组中的元素而不改变其余元素的顺序。
例如，[3,6,2,7] 是数组 [0,3,1,6,2,2,7] 的子序列。

 
示例 1：
输入：nums = [10,9,2,5,3,7,101,18]
输出：4
解释：最长递增子序列是 [2,3,7,101]，因此长度为 4 。

示例 2：
输入：nums = [0,1,0,3,2,3]
输出：4

示例 3：
输入：nums = [7,7,7,7,7,7,7]
输出：1

提示：
1 <= nums.length <= 2500
-104 <= nums[i] <= 10^4

*/

class Solution {
public:
    int lengthOfLIS(vector<int>& nums) {
        if (nums.size() <= 1)
            return nums.size();
        
        vector<int> dp(nums.size(), 1);
        int result = 0;
        for (int i = 1; i < nums.size(); i++) {
            for (int j = 0; j < i; j++) {
                if (nums[i] > nums[j])
                    dp[i] = max(dp[i], dp[j] + 1);
            }
            if (dp[i] > result)
                result = dp[i]; // 取长的子序列
        }
        return result;
    }
};

/* 参考代码随想录

最长上升子序列是动规的经典题目，这里dp[i]是可以根据dp[j]
（j < i）推导出来的，那么依然用动规五部曲来分析详细一波：

1. dp[i]的定义
dp[i]表示i之前包括i的以nums[i]结尾最长上升子序列的长度

2. 状态转移方程
位置i的最长升序子序列等于j从0到i-1各个位置的最长升序子序列 + 1 的最大值。

所以：if (nums[i] > nums[j]) dp[i] = max(dp[i], dp[j] + 1);

注意这里不是要dp[i] 与 dp[j] + 1进行比较，而是我们要取dp[j] + 1的最大值。

3. dp[i]的初始化
每一个i，对应的dp[i]（即最长上升子序列）起始大小至少都是1.

4. 确定遍历顺序
dp[i] 是有0到i-1各个位置的最长升序子序列 推导而来，
那么遍历i一定是从前向后遍历。

j其实就是0到i-1，遍历i的循环在外层，遍历j则在内层，代码如下：

for (int i = 1; i < nums.size(); i++) {
    for (int j = 0; j < i; j++) {
        if (nums[i] > nums[j]) dp[i] = max(dp[i], dp[j]+1);
    }
    if (dp[i] > result) result = dp[i]; // 取长的子序列
}

5. 举例推导dp数组

本题最关键的是要想到dp[i]由哪些状态可以推出来，
并取最大值，那么很自然就能想到递推公式：
dp[i] = max(dp[i], dp[j] + 1);

子序列问题是动态规划的一个重要系列，
本题算是入门题目，好戏刚刚开始！
*/